crates/energy/src/graph.rs

```diff
@@ -0,0 +1,239 @@
+    fn default() -> Self {
+        Self(Entity::PLACEHOLDER)
+    }
+}
+
+impl From<NearbyEntity> for Entity {
+    fn from(val: NearbyEntity) -> Self {
+        val.0
+    }
+}
+
+impl From<Entity> for NearbyEntity {
+    fn from(entity: Entity) -> Self {
+        Self(entity)
+    }
+}
+
+/// The power grid resource is used to store the power grid graph.
+#[derive(Resource, Debug, Clone)]
+pub(crate) struct PowerGrid {
+    /// The power grid graph.
+    graph: GraphMap<Entity, f64, Undirected>,
+}
+
+impl Default for PowerGrid {
+    fn default() -> Self {
+        Self {
+            graph: GraphMap::new(),
+        }
+    }
+}
+
+/// The energy grid member component is used to store the energy grid member entities.
+#[derive(Component, Debug, Clone)]
+pub struct EnergyGridMember;
+
+/// The nearby units component is used to store the nearby entities of an entity.
+#[derive(Component, Default, Debug, Clone)]
+pub struct NearbyUnits {
+    units: TinyVec<[NearbyEntity; 16]>,
+    last_pos: Option<Vec2>,
+}
+
+impl NearbyUnits {
+    fn remove_matching(&mut self, entity: NearbyEntity) {
+        self.units.retain(|e| *e != entity);
+        println!("removed {:?} from {:?}", entity, self.units);
+    }
+
+    pub fn len(&self) -> usize {
+        self.units.len()
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.units.is_empty()
+    }
+}
+
+fn setup(mut commands: Commands) {
+    commands.insert_resource(PowerGrid::default());
+}
+
+fn clean_up(mut commands: Commands) {
+    commands.remove_resource::<PowerGrid>();
+}
+
+/// This system spawns Energy Producers and Energy Receivers and nearby units.
+fn spawn_graph_components(
+    mut commands: Commands,
+    newly_spawned_units: Query<Entity, Added<Battery>>,
+) {
+    for entity in newly_spawned_units.iter() {
+        commands
+            .entity(entity)
+            .insert((EnergyGridMember, NearbyUnits::default()));
+    }
+}
+
+pub fn update_nearby(
+    spacial_index_member: SpatialQuery<Entity, With<EnergyGridMember>>,
+    mut nearby_units: Query<(Entity, &mut NearbyUnits, &Transform), Changed<Transform>>,
+) {
+    nearby_units
+        .par_iter_mut()
+        .for_each_mut(|(entity, mut nearby_units, transform)| {
+            let current_pos = transform.translation.to_flat();
+            if let Some(last_pos) = nearby_units.last_pos {
+                if current_pos.distance_squared(last_pos) < MIN_DISTANCE_FOR_UPDATE {
+                    return;
+                }
+            }
+            nearby_units.last_pos = Some(current_pos);
+
+            let aabb = &Aabb::new(
+                Point::from(transform.translation - Vec3::splat(MAX_DISTANCE)),
+                Point::from(transform.translation + Vec3::splat(MAX_DISTANCE)),
+            );
+
+            let members = spacial_index_member.query_aabb(aabb, Some(entity));
+
+            nearby_units.units.clear();
+
+            nearby_units.units.extend(members.map(NearbyEntity));
```

Comment:
This system updates only entities which moved by some minimal distance -> there must be reciprocal update of non-moving entities which got close / far from neighbors do the the movement of neighbors.

Possible solution:
* get difference between past and future nearby units,
* add self to all newly added units,
* remove self from all newly removed units,
* update units on self.